Small improvements following scratchpad feature.

r0xsh opened the pull request:
- Introduce a special variable `_` that will be automatically set to the result of the preceding line.
- Remove the `q` keybind since you seems to prefer to not set them at plugin level
- Introduce a config key `scratchpad.virtual_text_format` (so i can use a double space :smiley:)

r0xsh committed:
feat(scratchpad): create a special variable `_` set to the previous eval

r0xsh committed:
refact(scratchpad): remove keybind `q`.

r0xsh committed:
feat(scratchpad): allow to set `virtual_text_format` for the eval virtual text

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -56,7 +56,7 @@ end
 local function render_result(line_idx, text, col)
 	vim.api.nvim_buf_set_extmark(state.buf, state.ns, line_idx, col, {
-		virt_text = { { " = " .. text, "Comment" } },
+		virt_text = { { string.format(config.options.scratchpad.virtual_text_format, text), "Comment" } },
```
Add `"Comment"` as a default option aswell, so people can change the highlight. It would even be better if people could choose a different color for their ` = ` and for their `text` (that you should rename to "answer" or "result"), although the default will be "Comment" for the whole virtual text.

r0xsh replied:
> Add "Comment" as a default option as well

This seems reasonable (:

> It would even be better if people could choose a different color for their = and for their text 

I mean i see the appeal for this but maybe avoid going into a hyper configurable rabbit hole + it require some string parsing/manipulation to achieve this or directly expose the `virt_text` table to the end user, to which i'm not a huge fan.
```lua
vim.api.nvim_buf_set_extmark(state.buf, state.ns, line_idx, col, {
		virt_text = config.options.scratchpad.i_suck_at_naming_things,
		...
```

Necrom4 replied:
Personally I don't like exposing the whole table to the end user either, I just previously had someone ask me something similar and thought it may be something as of `calcium` in the future.

For now let's leave it at `format` + `highlight` options only then.

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -72,13 +72,16 @@ function M.evaluate_scratchpad()
 
 	for i, line in ipairs(lines) do
 		if line ~= "" then
+			variables["_"] = previous_evaluation
```
Rename to `ans` to follow `qalc` (cli-tool) standard. If you believe `_` is better, implement both.
```suggestion
			variables["ans"] = previous_evaluation
```

r0xsh replied:
I think `ans` is more standard here :+1: 
But seems very simple to make this configurable too

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -92,10 +95,6 @@ function M.create_scratchpad()
 
 	state.buf, state.win = setup_window()
 
-	-- Keymaps
-	local map_opts = { buffer = state.buf, silent = true }
-	vim.keymap.set("n", "q", M.close_scratchpad, map_opts)
-
```
While I did prefer to remove the other `<c-l>` keymap, this one has become such a standard in neovim plugins that create a floating window that I believe we should keep it.

r0xsh replied:
Huh, in this case. `<Esc>` is also kinda a standard, should i introduce it or only keep the `q` one

Necrom4 replied:
Honestly, from trying out my plugins it's pretty 50/50, for example folke's `Snacks.scratch()` only closes with `q`. In `calcium`'s case, I believe we should have leave `<esc>` be as this is a buffer people will edit in in multiple modes.
For example if they leave insert mode (one click) and misclick with a second click, the buffer will close and they'll loose everything they were working on (as your implementation of scratchpad clears after quitting (we could change that btw)).

 Let's keep `q` as a default and make it configurable for the user.

r0xsh replied:
Make sense, let leave it as it is.

> (as your implementation of scratchpad clears after quitting (we could change that btw))

Yup totally agree on that one, i thought implementing this also

Necrom4 commented on README.md:
```diff
@@ -32,6 +32,10 @@ A powerful [`lua-lib-math`](https://www.lua.org/pil/18.html) in-buffer calculato
+	scratchpad = {
+		border = "rounded", -- scratchpad border style, :help 'winborder' 
+		virtual_text_format = " = %s", -- scratchpad virtual text format
+	},
```
Format this so that it conforms with Markdown linters, these 4 lines are using hard tabs and hence are not aligned with the rest of the text.

Necrom4 commented on lua/calcium/config.lua:
```diff
@@ -5,6 +5,7 @@ M.defaults = {
 	default_mode = "append",
 	scratchpad = {
 		border = "rounded",
+		virtual_text_format = " = %s",
```
As mentioned in the other comment, it'd be better to have more options for the virtual text than just a string.

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -41,7 +41,7 @@ local function setup_window()
 		height = height,
 		row = math.floor((ui.height - height) / 2),
 		col = math.floor((ui.width - width) / 2),
-		border = config.options.scratchpad.border or "rounded",
```
Just curious, why did you remove this?

r0xsh replied:
It was an oversight. Before introducing the configs for the scratchpad, I used `win.o.winborder or "rounded"`. But since the fallback is done when we merge the default config and user config together (in config module), there is no need to also have a fallback downstream.

Necrom4 commented on README.md:
```diff
@@ -32,6 +32,10 @@ A powerful [`lua-lib-math`](https://www.lua.org/pil/18.html) in-buffer calculato
     default_mode = "append", -- or `replace` the expression
+	scratchpad = {
+		border = "rounded", -- scratchpad border style, :help 'winborder' 
+		virtual_text_format = " = %s", -- scratchpad virtual text format
```
No need to repeat `scratchpad` while inside the scratchpad scope
Also if you're going to name it "virtual text" here, rename every mention of "Ghost Text" everywhere else to be in accordance with this. (I do believe "virtual text" is indeed better).
```suggestion
		border = "rounded", -- floating window border style, :help 'winborder' 
		virtual_text_format = " = %s", -- virtual text format
```

Necrom4 reviewed (changes requested):
Great work mate! Can't believe you had time to create another PR while I went sleeping.
Again I wrote everything as code diff comments.
Lastly, please add everything to the README, including an explanation for `ans`/`_`.

r0xsh commented:
Yeah, I ended up figuring out that introducing the `ans`/`_` behavior was way easier than anticipated. Couldn't resist, haha. 
I'll take care of your feedback this weekend (:

Necrom4 commented:
Perfect, I implemented syntax highlighting for the scratchpad while you were away, I probably didn't follow the latest modern standards (if there are any) as I was focused on getting those damn Regex patterns to work, haha. If you have any feedback, don't hesitate to comment on the (now) closed PR.
See you in the weekend!

r0xsh committed:
feat(scratchpad): extend scratchpad config.

```lua
scratchpad = {
        border = "rounded",
        virtual_text = {
                format = " = %s",
                highlight_group = "Comment",
        },
result_variable = "ans"
}
```

r0xsh committed:
Revert "refact(scratchpad): remove keybind `q`."

This reverts commit 343b5ba8c09187828f5f23f70c8d5f37c0ddc0d7.

r0xsh committed:
docs(scratchpad): add exemples in README.md

Necrom4 commented on lua/calcium/config.lua:
```diff
@@ -5,6 +5,11 @@ M.defaults = {
 	scratchpad = {
 		border = "rounded",
+		virtual_text = {
+			format = " = %s",
```
Remove the space due to changes done during the syntax highlighting PR, see 284c4ce9f4c9fe9a5a85d11063e67fa590f5e0a2
```suggestion
			format = "= %s",
```

Necrom4 commented on README.md:
```diff
@@ -30,8 +30,16 @@ A powerful [`lua-lib-math`](https://www.lua.org/pil/18.html) in-buffer calculato
+    scratchpad = {
+        border = "rounded",               -- floating window border style, :help 'winborder' 
+        virtual_text = {
+            format = " = %s",             -- virtual text format
```
Same as before.
```suggestion
            format = "= %s",              -- virtual text format
```

Necrom4 reviewed (changes requested):
Great work mate! Just a few last changes.

r0xsh commented:
I think I covered all the feedback. Let me know if something is not okay. 
I'll change the "clear on scratchpad close" in another PR :)

r0xsh committed:
fix(scratchpad): remove the space

See: https://github.com/Necrom4/calcium.nvim/commit/284c4ce9f4c9fe9a5a85d11063e67fa590f5e0a2

Necrom4 commented:
All good, great work mate! 🎉
